### DISCUSSION TITLE
NodeJS compatible API?

### DISCUSSION THREAD

**Discussion by med-ab** [c-smile/sciter-js-sdk#275] [General]
It says in the readme that:
> Node.JS runtime (more or less full) is coming;

How compatible is sciter with node and how long it might take for somewhat full compatibility like [neonious/lowjs](/neonious/lowjs)

**Comment by c-smile** [comment]
Full compatibility is unclear. 

If you mean built-in modules then it is to be decided what is needed beyond of these: https://github.com/c-smile/sciter-js-sdk/blob/main/docs/md/module-sys.md 

What of [built-in NodeJS](https://www.w3schools.com/nodejs/ref_modules.asp  ) features are required in Sciter context? 

And there is also a question about third-party NodeJS modules. Problem is that they use V8 API directly. And that one is not available in Sciter that use QuickJS instead.

Sciter `@sys` module is close by feature set to NodeJS API but not exact. In principle it is possible to make a polyfill to achieve 100% compatibility with chosen modules.

**Comment by med-ab** [comment]
Well, for what is needed beyond the current built-in sys modules, i think the following node built-ins would be very useful:
[`https.createServer()`](https://nodejs.org/api/https.html#httpscreateserveroptions-requestlistener)
[`path.join()`](https://nodejs.org/api/path.html#pathjoinpaths) and [`path.sep()`](https://nodejs.org/api/path.html#pathsep)
[`dns.lookup()`](https://nodejs.org/api/dns.html#dnslookuphostname-options-callback) and [`dns.resolve()`](https://nodejs.org/api/dns.html#dnsresolvehostname-rrtype-callback)
[`os.arch()`](https://nodejs.org/api/os.html#osarch), [`os.cpus()`](https://nodejs.org/api/os.html#oscpus), [`os.freemem()`](https://nodejs.org/api/os.html#osfreemem), [`os.totalmem()`](https://nodejs.org/api/os.html#ostotalmem), [`os.hostname()`](https://nodejs.org/api/os.html#oshostname), [`os.uptime()`](https://nodejs.org/api/os.html#osuptime) and [`os.version()`](https://nodejs.org/api/os.html#osversion)
[`process.argv`](https://nodejs.org/api/process.html#processargv), [`process.versions`](https://nodejs.org/api/process.html#processversions), [`process.uptime`](https://nodejs.org/api/process.html#processuptime), [`process.memoryUsage()`](https://nodejs.org/api/process.html#processmemoryusage) and [`process.cpuUsage()`](https://nodejs.org/api/process.html#processcpuusagepreviousvalue)

For third-party modules, i suppose you mean javascript es modules that are usually downloadable through npm. They generally use [nodejs API](https://nodejs.org/api/) and javascript standard objects and statements.. not V8 API directly (though very few do but v8 is only one api in the list)

**Reply by c-smile** [reply]
That's clearly server side set of functions. Not that frequently used on clients.
It is not that difficult to provide them, it just that no one requested them so far. But is possible, of course.

Strange, I thought node-api initiative did not go through. AFAIR it was pushed by MS when they tried to replace V8 by Chakra, but I haven't heard about success of that last years.

**Reply by med-ab** [reply]
Would implementing these functions be in the schedule or i should discuss it in the txiki project?

Oh you mean [node-api](https://nodejs.org/api/n-api.html) (N-API) for C/C++ addons? i was talking about the nodejs api, like the standard library of builtin modules (fs, http, console...)
I don't know much about C/C++ TBH, im just a JS dev.
